[Merged by Bors] - feat: Add some oangle theorems

wangying11123 opened the pull request:
Add several theorems relating oriented angles and affine independence / collinearity, together with auxiliary lemmas about the equality and sign of oriented angles.

Main additions

`oangle_ne_zero_and_ne_pi_iff_not_collinear`
– An oriented angle is neither `0` nor `π` if and only if the three points are not collinear.

`affineIndependent_iff_of_oangle_eq`
– If two oriented angles are equal, then the corresponding triples of points are affinely independent simultaneously.

`not_collinear_iff_of_angle_eq`
– If two oriented angles are equal, one triple is collinear if and only if the other is.

`oangle_eq_or_eq_neg_of_angle_eq`
– The oriented angles are equal or opposite if the unoriented angles are equal.

`angle_eq_neg_of_two_zsmul_angle_eq`
– If `2 • a = 2 • b` and the signs are opposite, then `a = b + π.`

`two_zsmul_eq_iff_eq`
– Characterizes equality of doubled angles when signs are equal.

`oangle_sub_oangle_ne_pi_of_not_collinear`
– If the signs of two oriented angles are equal and the points are not collinear, their difference is not `π`.

wywangying-wy committed:
add Exterior angle theorem

wywangying-wy committed:
Merge remote-tracking branch 'upstream/master' into dsfxcimk

wywangying-wy committed:
conflict solved

wangying11123 committed:
Merge branch 'leanprover-community:master' into master

wangying11123 committed:
Merge branch 'leanprover-community:master' into master

Commit:
add oangle_ne_zero_and_ne_pi_iff_not_collinear and oangle_eq_or_eq_neg_of_angle_eq

Commit:
add some theorem about oriented angle

wangying11123 committed:
Remove exterior angle theorem from Triangle.lean

Remove a theorem that was already submitted in a previous PR.

Commit:
change fenzhi

Commit:
commit oangle

wangying11123 committed:
Update Between.lean

delete my wrong change

wangying11123 committed:
Update Affine.lean

Delete some redundant blank lines

wangying11123 committed:
Merge branch 'master' into yulia

dsfxcimk commented on Mathlib/Geometry/Euclidean/Angle/Oriented/Affine.lean:
```diff
@@ -200,6 +201,27 @@ theorem oangle_sign_eq_zero_iff_collinear {p₁ p₂ p₃ : P} :
+/-- An oriented angle is not zero and `π` if and only if the three points are not collinear. -/
+theorem oangle_ne_zero_and_ne_pi_iff_not_collinear {p₁ p₂ p₃ : P} :
+  ∡ p₁ p₂ p₃ ≠ 0 ∧ ∡ p₁ p₂ p₃ ≠ π ↔ ¬ Collinear ℝ {p₁, p₂, p₃} := by
+  rw[oangle_ne_zero_and_ne_pi_iff_affineIndependent, affineIndependent_iff_not_collinear_set]
```
Thanks! This has now been fixed.

dsfxcimk commented on Mathlib/Geometry/Euclidean/Angle/Oriented/Affine.lean:
```diff
@@ -333,6 +355,67 @@ theorem angle_eq_iff_oangle_eq_of_sign_eq {p₁ p₂ p₃ p₄ p₅ p₆ : P} (h
+    aesop
+
+/-- If the signs of two oriented angles are equal and triple of points is not collinear,
+then their difference is not π when the second. -/
```
Thanks! This has now been fixed.

jsm28 commented on Mathlib/Analysis/SpecialFunctions/Trigonometric/Angle.lean:
```diff
@@ -890,6 +890,31 @@ theorem sign_two_zsmul_eq_sign_iff {θ : Angle} :
   rw [two_zsmul, ← two_nsmul, sign_two_nsmul_eq_sign_iff]
 
+theorem angle_eq_neg_of_two_zsmul_angle_eq (a b : Real.Angle) (h : 2 • a = 2 • b)
+  (h_sign : a.sign = -b.sign) (h_ne : b.sign ≠ 0) : a = b + π := by
```
```suggestion
theorem eq_add_pi_of_two_zsmul_eq_of_sign_eq_neg (a b : Real.Angle) (h : 2 • a = 2 • b)
  (h_sign : a.sign = -b.sign) (h_ne : b.sign ≠ 0) : a = b + π := by
```

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Oriented/Affine.lean:
```diff
@@ -212,6 +212,19 @@ theorem oangle_sign_eq_zero_iff_collinear {p₁ p₂ p₃ : P} :
+  (h_eq : ∡ p₁ p₂ p₃ = ∡ p₁' p₂' p₃') :
+  AffineIndependent ℝ ![p₁, p₂, p₃] ↔ AffineIndependent ℝ ![p₁', p₂', p₃'] := by
+  rw [←oangle_ne_zero_and_ne_pi_iff_affineIndependent,
+  ←oangle_ne_zero_and_ne_pi_iff_affineIndependent, h_eq]
```
```suggestion
  rw [← oangle_ne_zero_and_ne_pi_iff_affineIndependent,
    ← oangle_ne_zero_and_ne_pi_iff_affineIndependent, h_eq]
```

wangying11123 committed:
Update Mathlib/Analysis/SpecialFunctions/Trigonometric/Angle.lean

rename angle_eq_neg_of_two_zsmul_angle_eq as eq_add_pi_of_two_zsmul_eq_of_sign_eq_neg

Co-authored-by: Joseph Myers <jsm@polyomino.org.uk>

wangying11123 committed:
Update Mathlib/Analysis/SpecialFunctions/Trigonometric/Angle.lean

Co-authored-by: Joseph Myers <jsm@polyomino.org.uk>

wangying11123 committed:
Update Mathlib/Geometry/Euclidean/Angle/Oriented/Affine.lean

Co-authored-by: Joseph Myers <jsm@polyomino.org.uk>

wangying11123 committed:
Update Mathlib/Analysis/SpecialFunctions/Trigonometric/Angle.lean

Co-authored-by: Joseph Myers <jsm@polyomino.org.uk>

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Oriented/Affine.lean:
```diff
@@ -212,6 +212,19 @@ theorem oangle_sign_eq_zero_iff_collinear {p₁ p₂ p₃ : P} :
+/-- If two oriented angles are equal, one triple is affinely independent if and only if
+the other is. -/
+theorem affineIndependent_iff_of_oangle_eq {p₁ p₂ p₃ p₁' p₂' p₃' : P}
+  (h_eq : ∡ p₁ p₂ p₃ = ∡ p₁' p₂' p₃') :
```
Since we have the stronger lemma `affineIndependent_iff_of_two_zsmul_oangle_eq`, I'm doubtful of adding this variant as well. (If at the site of use you do have `h` saying the two angles are equal, not just equal mod pi, then you should be able to pass `h ▸ rfl` to that stronger lemma.)

wangying11123 committed:
Update Affine.lean

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Oriented/Affine.lean:
```diff
@@ -345,6 +350,41 @@ theorem angle_eq_iff_oangle_eq_of_sign_eq {p₁ p₂ p₃ p₄ p₅ p₆ : P} (h
+theorem oangle_sub_oangle_ne_pi_of_not_collinear {a b c a' b' c' : P}
+  (h_not_col' : ¬ Collinear ℝ ({a', b', c'} : Set P))
+  (h_sign : (∡ a b c).sign = (∡ a' b' c').sign) :
+  ∡ a b c - ∡ a' b' c' ≠ π := by
```
Rather than making this a very specific lemma about two triples of points and `Collinear`, I think it's more natural to have a lemma (in `Mathlib.Analysis.SpecialFunctions.Trigonometric.Angle`) about two `Real.Angle` values (if they have the same nonzero sign, the difference is not `π`), and then let users combine that with existing statements about `Collinear` as needed.

jsm28 replied:
I think that, having added that new lemma, this one still needs to be deleted.

That other PR is marked awaiting-author; if you think it's ready for review again, you should remove that label (comment with just "-awaiting-author" if you can't remove it directly).

wangying11123 committed:
Update Angle.lean

wangying11123 committed:
Delete oangle_sub_oangle_ne_pi_of_not_collinear

wangying11123 committed:
Fix theorem signature for two_zsmul_eq_of_sign_eq_neg

wangying11123 committed:
Fix theorem signatures and formatting in Angle.lean

wangying11123 committed:
Update Angle.lean

jcommelin reviewed (commented):
Thanks :tada:

bors merge

wangying11123 committed:
Merge branch 'leanprover-community:master' into yulia

Commit:
use smip only instend of simp
